cros_gralloc/cros_gralloc_driver.cc

```diff
@@ -230,151 +204,134 @@ cros_gralloc_driver::cros_gralloc_driver()
+				gpu_grp_type_idx = GPU_GRP_TYPE_INTEL_DGPU_IDX;
+			} else {
+				gpu_grp_type_idx = GPU_GRP_TYPE_INTEL_IGPU_IDX;
 			}
-		} else {
-			drv_video_ = drv_render_;
-			if (!drv_video_)
-				drv_loge("Failed to create driver for the video device with card id %d\n",
-					video_idx);
 		}
 
-		if ((virtio_node_idx != -1) && (virtio_node_idx != renderer_idx)) {
-			drv_kms_ = drv_create(node_fd[virtio_node_idx]);
-			if (!drv_kms_) {
-				drv_loge("Failed to create driver for the virtio-gpu device with card id %d\n",
-                                          virtio_node_idx);
-				close(node_fd[virtio_node_idx]);
-				drv_kms_ = drv_render_;
-                        }
-		} else {
-			drv_kms_ = drv_render_;
+		// Use the first device as fallback in case we don't have any known device.
+		if (drv_fallback_ == nullptr) {
+			drv_fallback_ = drv;
 		}
 
-		// Init drv
-		DRV_INIT(drv_render_, gpu_grp_type_, renderer_idx)
-		if (video_idx != renderer_idx)
-			DRV_INIT(drv_video_, gpu_grp_type_, video_idx)
-		if ((virtio_node_idx != -1) && (virtio_node_idx != renderer_idx))
-			DRV_INIT(drv_kms_, gpu_grp_type_, virtio_node_idx)
-		if (drv_kms_ && (virtio_node_idx != renderer_idx) && (drv_kms_ != drv_render_)) {
-			bool virtiopic_with_blob = drv_virtpci_with_blob(drv_kms_);
-			if (virtiopic_with_blob) {
-				//igpu SRIOV or dGPU SRIOV case
-				drv_logi("Virtio gpu device with blob\n");
-				if ((drv_kms_ != drv_render_) && drv_kms_)
-					DRV_DESTROY(drv_kms_)
-				drv_kms_ = drv_render_;
-			} else if (drv_is_dgpu(drv_video_) && drv_virtgpu_is_ivshm(drv_kms_)){
-				//is dgpu passthrough + ivshm case
-				drv_logi("dGPU with Virtio ivshm\n");
-				drv_ivshmem_ = drv_kms_;
-				drv_kms_ = drv_render_;
-				if (ivshm_node_idx != -1) {
-					close(node_fd[ivshm_node_idx]);
-					ivshm_node_idx = -1;
-				}
-			} else {
-				// is QNX or redhat case
-				drv_logi("Virtio ivshm or no blob\n");
-			}
-		}
-		if (ivshm_node_idx != -1) {
-			if (!(drv_ivshmem_ = drv_create(node_fd[ivshm_node_idx]))) {
-				drv_loge("Failed to create driver for the ivshm device with card id %d\n",
-					ivshm_node_idx);
-				close(node_fd[ivshm_node_idx]);
-			} else {
-				DRV_INIT(drv_ivshmem_, gpu_grp_type_, ivshm_node_idx)
-				if (drv_virtgpu_is_ivshm(drv_ivshmem_)) {
-					drv_logi("Node is virtio-ivishmem node");
-				} else {
-					drv_logi("Node is NOT virtio-ivishmem node");
-						DRV_DESTROY(drv_ivshmem_)
-				}
-			}
+		if (gpu_grp_type_idx != -1 &&
+		    !(gpu_grp_type_ & (1ull << gpu_grp_type_idx))) {
+			gpu_grp_type_ |= (1ull << gpu_grp_type_idx);
+			drivers_[gpu_grp_type_idx] = drv;
+		} else if (drv != drv_fallback_) {
+			DRV_DESTROY(drv);
 		}
+
+		drmFreeVersion(version);
 	}
 
-	for (int i = 0; i < availabe_node; i++) {
-		free(node_name[i]);
-		if ((i != renderer_idx) && (i != video_idx) && (i != virtio_node_idx) && (i != ivshm_node_idx)) {
-			close(node_fd[i]);
-		}
+	set_gpu_grp_type();
```

Comment:
The gpu graphic type should be set before driver create, otherwise the type is not set.
i think in dual gpu case, if use the igpu display port, the display is blur.

Reply:
> The gpu graphic type should be set before driver create, otherwise the type is not set. i think in dual gpu case, if use the igpu display port, the display is blur.

The driver object won't be used before complete of the constructor of cros_gralloc_driver, what we must ensure is to set gpu_grp_type before using the drivers. And I don't think we can get correct gpu_grp_type value without creating the drivers -- we have to query the supported features.

Reply:
As the 'gpu_grp_type' is needed in 'i915_add_combinations()'->'i915_init()', which is still in driver init stage.

Reply:
> As the 'gpu_grp_type' is needed in 'i915_add_combinations()'->'i915_init()', which is still in driver init stage.

Understood. I will refine these code and use another way to accomplish the same goal...

Reply:
Code updated.